docs/knowledgebase/runtime/benchmarking.md

```diff
@@ -0,0 +1,547 @@
+1,1245146,128021,1,4,1,2
+1,1238746,126051,1,4,1,2
+1,1206004,126391,1,4,1,2
+1,1212564,127941,1,4,1,2
+100,1257646,129750,1,4,1,2
+100,1232476,125780,1,4,1,2
+100,1215466,128310,1,4,1,2
+100,1205835,129070,1,4,1,2
+...
+991,1208294,125820,1,4,1,2
+991,1209305,126921,1,4,1,2
+991,1203275,125031,1,4,1,2
+991,1234855,124190,1,4,1,2
+991,1337136,125060,1,4,1,2
+
+Median Slopes Analysis
+========
+-- Extrinsic Time --
+
+Model:
+Time ~=     1231
+    + b        0
+              µs
+
+Reads = 1 + (0 * b)
+Writes = 1 + (0 * b)
+
+Min Squares Analysis
+========
+-- Extrinsic Time --
+
+Data points distribution:
+    b   mean µs  sigma µs       %
+    1      1230     12.61    1.0%
+  100      1230     16.53    1.3%
+  199      1227     12.88    1.0%
+  298      1215     12.94    1.0%
+  397      1239     19.83    1.6%
+  496      1234     18.51    1.5%
+  595      1228     7.167    0.5%
+  694      1233     13.51    1.0%
+  793      1218     9.876    0.8%
+  892      1225     14.15    1.1%
+  991      1220     11.63    0.9%
+
+Quality and confidence:
+param     error
+b         0.006
+
+Model:
+Time ~=     1230
+    + b        0
+              µs
+
+Reads = 1 + (0 * b)
+Writes = 1 + (0 * b)
+
+// -- snipped on the second benchmark case
+```
+
+With the median slopes analysis, this is the weight function:
+
+```
+1231 µs + (0 * b) +
+  [1 + (0 * b)] * db read time +
+  [1 + (0 * b)] * db write time
+```
+
+We separate the db/storage read and write out because they are particularly expensive and their
+respective operation time will be retrieved from the runtime. We just need to measure how many db
+read write are performed with respect to the change of `b`.
+
+The benchmark result is telling us that it will always perform 1 db read and 1 db write no matter
+how `b` changes.
+
+The benchmark library gives us both a median slope analysis; that the execution time of a particular
+`b` value is taken as the median value of the repeated runs, and a [min. square
+analysis](https://en.wikipedia.org/wiki/Least_squares) that is better explained in a statistics
+primer.
+
+You can also derive your own coefficients given you have the raw data on each run, say maybe you
+know the computation time will not be a linear but an *O(nlogn)* relationship with the input
+variable. So you need to determine the coefficient differently.
+
+### Auto-generated `WeightInfo` Implementation
+
+The second output is an autogenerated `WeightInfo` implementation. This file defines weight
+functions of benchmarked extrinsics with the computed coefficient above. You can directly integrated
+this file in your extrinsics weight function or further customize them if so desired. The
+auto-generated implementation is designed to make end-to-end benchmark automation easy.
+
+To use this file, you first define a `WeightInfo` trait in
+[`frame/example/src/lib.rs`](https://github.com/paritytech/substrate/blob/master/frame/example/src/lib.rs):
+
+```rust
+pub trait WeightInfo {
```

Comment:
hmm wait -- WeightInfo is also generated in as the output.

Reply:
I believe by default when running benchmark as is, it doesn't generate the trait, but just the implementation. It is for Substrate now because the CI/developer is using [frame-weight-template.hbs](https://github.com/paritytech/substrate/blob/v3.0.0/.maintain/frame-weight-template.hbs) as the template to generate the `weightInfo` implementation by passing additional `--template` cli argument. 

I am not covering WeightInfo template here and just introducing the default generated `WeightInfo` impl output. 

@shawntabrizi Just want to confirm if my understanding is correct.

Reply:
You are correct but my question is why not cover this? I personally have found it a bit of pain to manually maintain correct signatures in `Trait WeightInfo` and the fact that it can be autogenerated, IMO, is important.